### ISSUE TITLE
netcore : ReloadMetadataIfRequired issue with local metadata file

### ISSUE THREAD

**Issue by radderz** [Sustainsys/Saml2#1110]
We just had an website outage caused by a case where the in the pipeline ReloadMetadataIfRequired  is called, which results in a reload of the metadata file. We are still diagnosing how that could occur. We are getting an System.Net.WebException of operation has timed out.

The Metadata file is stored in App_Data which is locally accessible only, technically a webclient can access that, but one option would be to check if the metadataLocation is local file first and if it is to just read that file directly.

```
System.Net.WebException:
   at System.Net.FileWebRequest.GetResponse (System.Net.Requests, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Net.WebClient.GetWebResponse (System.Net.WebClient, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at System.Net.WebClient.OpenRead (System.Net.WebClient, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at Sustainsys.Saml2.Metadata.MetadataLoader.Load (Sustainsys.Saml2, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null)
   at Sustainsys.Saml2.Metadata.MetadataLoader.LoadIdp (Sustainsys.Saml2, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null)
   at Sustainsys.Saml2.IdentityProvider.DoLoadMetadata (Sustainsys.Saml2, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null)
   at Sustainsys.Saml2.IdentityProvider.ReloadMetadataIfRequired (Sustainsys.Saml2, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null)
   at ViewerCore.Startup.<ConfigureServices>b__3_10 (ViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: C:\Git\BlackhawkPlatform\ViewerCore\Startup.csViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 157)
   at Microsoft.Extensions.Options.OptionsFactory`1.Create (Microsoft.Extensions.Options, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Lazy`1.ViaFactory (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Lazy`1.CreateValue (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Sustainsys.Saml2.AspNetCore2.Saml2Handler.InitializeAsync (Sustainsys.Saml2.AspNetCore2, Version=2.2.0.0, Culture=neutral, PublicKeyToken=null)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandlerProvider+<GetHandlerAsync>d__5.MoveNext (Microsoft.AspNetCore.Authentication.Core, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware+<Invoke>d__6.MoveNext (Microsoft.AspNetCore.Authentication, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.ResponseCaching.ResponseCachingMiddleware+<Invoke>d__9.MoveNext (Microsoft.AspNetCore.ResponseCaching, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware+<Invoke>d__7.MoveNext (Microsoft.AspNetCore.StaticFiles, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware+<Invoke>d__6.MoveNext (Microsoft.AspNetCore.Diagnostics, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware+<Invoke>d__6.MoveNext (Microsoft.AspNetCore.Diagnostics, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at ViewerCore.Middleware.ReverseProxyMiddleware+<Invoke>d__3.MoveNext (ViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: C:\Git\BlackhawkPlatform\ViewerCore\Middleware\ReverseProxyMiddleware.csViewerCore, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 37)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1+<ProcessRequestAsync>d__2.MoveNext (Microsoft.AspNetCore.Server.IIS, Version=2.2.6.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)

```

One option might be to do this if the error above doesn't have a different known cause (i.e. just open the file as a bytes):

```
private static MetadataBase Load(
            string metadataLocation,
            IEnumerable<SecurityKeyIdentifierClause> signingKeys,
            bool validateCertificate,
            string minIncomingSigningAlgorithm)
        {
            XmlDictionaryReader reader;
            if (System.IO.File.Exists(metadataLocation))
            {
                var bytes = System.IO.File.ReadAllBytes(metadataLocation);

                reader = XmlDictionaryReader.CreateTextReader(
                    bytes,
                    XmlDictionaryReaderQuotas.Max);
            }
            else
            {
                if (PathHelper.IsWebRootRelative(metadataLocation))
                {
                    metadataLocation = PathHelper.MapPath(metadataLocation);
                }

                using (var client = new WebClient())
                using (var stream = client.OpenRead(metadataLocation))
                using (var ms = new MemoryStream())
                {
                    byte[] buf = new byte[65536];
                    for (; ; )
                    {
                        int read = stream.Read(buf, 0, buf.Length);
                        if (read == 0)
                            break;
                        ms.Write(buf, 0, read);
                    }
                    // System.Diagnostics.Debug.WriteLine(Encoding.UTF8.GetString(ms.ToArray()));
                    ms.Position = 0;
                    reader = XmlDictionaryReader.CreateTextReader(
                        ms,
                        XmlDictionaryReaderQuotas.Max);
                }
            }

            if (signingKeys != null)
            {
                reader = ValidateSignature(
                    reader,
                    signingKeys,
                    validateCertificate,
                    minIncomingSigningAlgorithm);
            }

            return Load(reader);            
        }
```

**Comment by AndersAbel** [comment]
The entire metadata loading need a redesign to allow more flexibility and better error handling/resiliency. This is just another proof of that.

**Comment by radderz** [comment]
> The entire metadata loading need a redesign to allow more flexibility and better error handling/resiliency. This is just another proof of that.

The problem with this error is it brings the whole site down as it is part of the pipeline. It was causing this to become an error even for requests that had nothing to do with SAML.

**Comment by AndersAbel** [comment]
> > The entire metadata loading need a redesign to allow more flexibility and better error handling/resiliency. This is just another proof of that.
> 
> The problem with this error is it brings the whole site down as it is part of the pipeline. It was causing this to become an error even for requests that had nothing to do with SAML.

Ok, didn't understand that first. Then it's worse. The SAML2 library should of course only be activated if the request is SAML2-related. Looking at the stack trace, I think the issue is that the options are accessed by the Saml2Handler before the check if "is this a SAML2 request". It should of course be the other way around.

Would that be a reasonable first fix for you? To ensure that any metadata load failure does not cause issues on other requests?

**Comment by radderz** [comment]
Yea thatd be a much better situation

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by radderz** [comment]
Should I to fix this and do a PR or do you know where it happens? Is it likely to be in the next release? It's having a really big impact even for users not using SAML based Auth when the SAML system gets stuck into a bad state. The only option when it gets there is to do a recycle.

**Comment by radderz** [comment]
could it have anything to do with the order of what I am doing in the startup.cs?

```
.AddSaml2(options =>
{
    options.SPOptions = CreateSPOptions();

    var idp = new IdentityProvider(new EntityId("https://otherparty/adfs/services/trust"), options.SPOptions)
    {
        //LoadMetadata = true,
        AllowUnsolicitedAuthnResponse = true,
        Binding = Saml2BindingType.HttpRedirect,
        SingleSignOnServiceUrl = new Uri("https://otherparty/adfs/ls/"), //The url where the identity provider listens for incoming sign on requests
    };

    idp.MetadataLocation = Path.Combine(_hostingEnvironment.ContentRootPath, "App_Data/Certificates/FederationMetadata2018.xml");

    idp.SigningKeys.AddConfiguredKey(new X509Certificate2(Path.Combine(_hostingEnvironment.ContentRootPath, "App_Data/Certificates/Signing2018.cer")));

    options.IdentityProviders.Add(idp);
    options.SignInScheme = "ExternalSaml";
});
```


I have also noticed in the identity provider code the locks are within the checks to see if it should do a load, 

A pattern more like would mean that it wouldn't double up on work as if multiple threads are waiting on the lock, they shouldn't need to load the metadata if the previous lock holder has done the work already: 

```
if (LoadMetadata && MetadataValidUntil.Value < DateTime.UtcNow)
{
    lock (metadataLoadLock)
    {
        if (LoadMetadata && MetadataValidUntil.Value < DateTime.UtcNow)
        {
            DoLoadMetadata();
        }
    }
}
```

**Comment by AndersAbel** [comment]
I've looked a bit more at this and my initial guess was wrong.

Looking at the stack trace I think that this is what is happening.
1. A request comes in and the Saml2 handler needs to check if it is a path mapped to Saml2.
2. The Saml2 handler checks the options to get the ModulePath (defaults to /Saml2, but can be configured to anything)
3. The options are lazily constructed when first accessed.
4. The configuration block in your code is run as part of option construction.
5. The configuration block creates and IdentityProvider object, which tries to load Metadata on creation.
6. The Metadata loading fails causing an exception.
7. Everything blows up because the configuration block is not supposed to throw exceptions.

If I remember correctly, I've written the code so that an exception from the initial metadata load at construction is bubbled up, while exceptions on refreshes just causes a retry. I think I had some intention that this way any incorrect config would be noticed and handled. But now it looks like a bad idea. I think that even during construction, a failed Metadata load should cause retries and an error being logged.

If you would like to look into this and create a PR that would be welcome. Please feel free to comment here with design thoughts. The MetadataLoader is a bit fragile (and the tests are horribly fragile) so it's not the easiest part of the code to work with.

**Comment by radderz** [comment]
@AndersAble, I think what I mentioned above for local file location, or a specified stream would work well. In the case of our file, we had no cache validity or valid till, as its just a permanent file on the machine rather than one available online. I added a valid till, which will atleast reduce reloading requirements but the initial reading of the file using a webrequest instead of just an File.ReadAllBytes or a stream provided is a hole in the design.

I have created a version that passes tests where it checks if the file exists first (i.e. the web app has permission to access the file) and just reads it directly as bytes if its a file instead of a URL. This should get rid of any possible issue on first load based on the webrequest style. The pain is it doesn't happen everytime, however if the website is being hammered just as its trying to initialise, I think it creates a deadlock situation or a continuous re/read due to the locks not being outside of the check if it needs to get a new version of the metadata.

I'll create a PR for you to check out the suggested changes.
